### PR TITLE
Support metric_relabel_configs in distributor

### LINKED ISSUE #1507: Consider allowing metric relabelling (or just "drop") during ingest
I think a nice feature would be to provide a way to use metric relabelling in the distributors.

The main use case I want is to be able to turn off certain sources of metrics quickly without having to change the remote write config of the prometheus instances running in each of my clusters.

### PR DESCRIPTION
**Pull request by roidelapluie**
Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

**What this PR does**:
Support metric_relabel_configs in distributor

**Which issue(s) this PR fixes**:
Fixes #1507 

**Checklist**
- [X] Tests updated
- [X] Documentation added
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by csmarchbanks — COMMENTED** [review]
:partying_face:, I am excited for this!

Do you think it is worthwhile to allow this to be specified per user rather than just globally? I imagine a common use case will be to stop very high cardinality or high churn data from a specific tenant without disrupting data for all tenants.

File: CHANGELOG.md [hunk 1]
```diff
@@ -49,6 +49,7 @@
 * [CHANGE] Increased default `-<prefix>.redis.timeout` from `100ms` to `500ms`. #3301
 * [FEATURE] Added support for shuffle-sharding queriers in the query-frontend. When configured (`-frontend.max-queriers-per-tenant` globally, or using per-tenant limit `max_queriers_per_tenant`), each tenants's requests will be handled by different set of queriers. #3113 #3257
 * [FEATURE] Query-frontend: added `compression` config to support results cache with compression. #3217
+* [FEATURE] Support `metric_relabel_confis` in distributor. #3329
```
Comment (drequena) [inline CHANGELOG.md:52]: configs instead of confis (just a simples typo) ;)

File: pkg/distributor/distributor.go [hunk 2]
```diff
@@ -170,6 +171,9 @@ type Config struct {
 	// when true the distributor does not validate the label name, Cortex doesn't directly use
 	// this (and should never use it) but this feature is used by other projects built on top of it
 	SkipLabelNameValidation bool `yaml:"-"`
+
+	// Metrics relabeling
+	MetricRelabelConfigs []*relabel.Config `yaml:"metric_relabel_configs,omitempty" doc:"nocli|description=List of metric relabel configurations. Note that in most situations, it is more effective to use metrics relabeling directly in the Prometheus server, e.g. remote_write.write_relabel_configs."`
```
Comment (pstibrany) [inline pkg/distributor/distributor.go:176]: We should note that this is global, for ALL tenants.

**Review by pstibrany — COMMENTED** [review]
Can we move the configuration into [Overrides](https://github.com/cortexproject/cortex/blob/c0d39957f7e509600773d028913ad765698538e6/pkg/util/validation/limits.go)? Then it can be per-user, and reloaded when changed on disk, without restarting Cortex. WDYT?

**Comment by roidelapluie** [comment]
I was already thinking about https://github.com/cortexproject/cortex/issues/3318#issuecomment-706976139 when I set this up. WDYT? Should we have both?

**Comment by pstibrany** [comment]
> I was already thinking about [#3318 (comment)](https://github.com/cortexproject/cortex/issues/3318#issuecomment-706976139) when I set this up. WDYT? Should we have both?

If by "both" you refer to global and per-tenant config, then Overrides / Limits already support that, and yes, I think it makes sense to have both. For example in cloud offering, we wouldn't use global one, but could use per-tenant config in specific cases. On the other hand, some people running Cortex internally may only need global one.

**Comment by bboreham** [comment]
Thanks!  Any idea what the CPU/memory overhead is?

**Comment by roidelapluie** [comment]
That depends on the relabeling rules..

**Comment by roidelapluie** [comment]
I have moved it to limits

File: pkg/distributor/distributor.go [hunk 3]
```diff
@@ -455,6 +456,11 @@ func (d *Distributor) Push(ctx context.Context, req *client.WriteRequest) (*clie
 			removeLabel(labelName, &ts.Labels)
 		}
 
+		if mrc := d.limits.MetricRelabelConfigs(userID); len(mrc) > 0 {
```
Comment (jtlisi) [inline pkg/distributor/distributor.go:459]: **question**: Is there any instrumentation or debug logging that could be added that would be helpful for operators/users when working with this feature?

**Review by jtlisi — APPROVED** [review]
LGTM!

File: CHANGELOG.md [hunk 4]
```diff
@@ -49,6 +49,7 @@
 * [CHANGE] Increased default `-<prefix>.redis.timeout` from `100ms` to `500ms`. #3301
 * [FEATURE] Added support for shuffle-sharding queriers in the query-frontend. When configured (`-frontend.max-queriers-per-tenant` globally, or using per-tenant limit `max_queriers_per_tenant`), each tenants's requests will be handled by different set of queriers. #3113 #3257
 * [FEATURE] Query-frontend: added `compression` config to support results cache with compression. #3217
+* [FEATURE] Support `metric_relabel_configs` in distributor. #3329
```
Comment (jtlisi) [inline CHANGELOG.md:52]:
```suggestion
* [FEATURE] Added the support for applying a Prometheus metrics relabel config on series received by the distributor. The `metric_relabel_configs` field was added to the limits config. #3329
```

File: pkg/distributor/distributor.go (hunk 3 above)
Comment (roidelapluie) [reply pkg/distributor/distributor.go]: distributor_received_samples_total and distributor_samples_in_total covers that.

File: CHANGELOG.md (hunk 4 above)
Comment (roidelapluie) [reply CHANGELOG.md]: Thanks, I updated my comment

**Comment by pstibrany** [comment]
Thank you, this looks very good! Would it be possible to add small test showing parsing of relabeling YAML rules into per-user limits?

**Comment by roidelapluie** [comment]
> Thank you, this looks very good! Would it be possible to add small test showing parsing of relabeling YAML rules into per-user limits?

Do we have tests like this already?

**Comment by pstibrany** [comment]
> > Thank you, this looks very good! Would it be possible to add small test showing parsing of relabeling YAML rules into per-user limits?
> 
> Do we have tests like this already?

Until now structures we've put into overrides YAML file were pretty basic. I can only see `TestLimitsLoadingFromYaml`. Something similar in size but with relabel config to make sure the wiring is all correct would be good to have. WDYT?

**Comment by roidelapluie** [comment]
Done :)

**Review by pracucci — COMMENTED** [review]
Could we mark it experimental in `docs/configuration/v1-guarantees.md`, please?

**Comment by roidelapluie** [comment]
done

File: docs/configuration/config-file-reference.template [hunk 5]
```diff
@@ -24,6 +24,7 @@ To specify which configuration file to load, pass the `-config.file` flag at the
 * `<string>`: a regular string
 * `<url>`: an URL
 * `<prefix>`: a CLI flag prefix based on the context (look at the parent configuration block to see which CLI flags prefix should be used)
+* `<relabel_config>`: a [prometheus relabeling configuration](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#relabel_config).
```
Comment (pracucci) [inline docs/configuration/config-file-reference.template:27]:
```suggestion
* `<relabel_config>`: a [Prometheus relabeling configuration](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#relabel_config).
```

**Review by pracucci — APPROVED** [review]
Thanks for working on this and sorry for the super late review (I would like days of 48h 😢 ). Looks flawless to me. I just left a nit, definitely not a blocker.

**Review by pstibrany — APPROVED** [review]
Thanks a lot Julien!

File: docs/configuration/config-file-reference.template (hunk 5 above)
Comment (pstibrany) [reply docs/configuration/config-file-reference.template]: Sorry, I've committed this thinking that I will fix DCO right after, but obviously I cannot push to your branch. 🤦

Comment (pstibrany) [reply docs/configuration/config-file-reference.template]: (Well, except via GitHub UI apparently)

Comment (roidelapluie) [reply docs/configuration/config-file-reference.template]: Yes you can

Comment (pstibrany) [reply docs/configuration/config-file-reference.template]: You're correct. I got confused there. Pushed DCO fix, when CI passes, I'll merge.

**Comment by bboreham** [comment]
I just used this feature in production to remove a high-cardinality metric; worked super well!

Even better if we had a metric to track how many samples got dropped this way - right now they disappear from `cortex_distributor_received_samples_total` and don't appear in `cortex_discarded_samples_total`.

**Pull request merged**